### PR TITLE
Clean up memory and spill contexts ownership passing

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

**Comment by findepi** [comment]
cc @dain

File: presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java [hunk 1]
```diff
@@ -140,7 +141,6 @@ private void writePages(Iterator<Page> pageIterator)
     public void close()
     {
         closer.register(() -> Files.delete(targetFileName));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java]: why not to put this to constructor as well?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 2]
```diff
@@ -139,7 +139,7 @@ public static ParquetPageSource createParquetPageSource(
             boolean predicatePushdownEnabled,
             TupleDomain<HiveColumnHandle> effectivePredicate)
     {
-        AggregatedMemoryContext systemMemoryContext = new AggregatedMemoryContext();
+        AggregatedMemoryContext systemMemoryUsage = new AggregatedMemoryContext();
```
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: If variable name now matches semantics then the class name seems to be invalid

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java [hunk 3]
```diff
@@ -149,15 +150,17 @@ public OrcReader(OrcDataSource orcDataSource, MetadataReader delegate, DataSize
             completeFooterSlice = Slices.wrappedBuffer(buffer, buffer.length - completeFooterSize, completeFooterSize);
         }
 
+        Supplier<LocalMemoryContext> untrackedMemory = new AggregatedMemoryContext().localContextSupplier();
```
Comment (kokosing) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: Maybe we could have something like `AggregateMemoryContext.untrackedMemory()`?

File: presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java (hunk 1 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java]: Unrelated. In fact I did this in spill PR

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 2 above)
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: Well.... can i ignore this comment? :)

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java (hunk 3 above)
Comment (findepi) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: The intention behind the variable name is to reveal the fact the memory seems to be accounted but the accounting goes to dev null (not reported anywhere). Any `new AggregateMemoryContext()` that is not read by anyone is untracked.

**Comment by losipiuk** [comment]
Could you please rebase before review?

**Comment by findepi** [comment]
@losipiuk, rebased

File: presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpillerFactory.java [hunk 4]
```diff
@@ -138,7 +139,7 @@ private static void cleanupOldSpillFiles(Path path)
     }
 
     @Override
-    public SingleStreamSpiller create(List<Type> types, SpillContext spillContext, LocalMemoryContext memoryContext)
+    public SingleStreamSpiller create(List<Type> types, Supplier<SpillContext> spillContext, LocalMemoryContext memoryContext)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpillerFactory.java]: I do not like it.

It makes the signature more complex. Actually it used to be somewhat similar to what you propose originally.
Maybe good enough solution if we are afraid that someone calls `close()` on passed `spillContext`, is to create new child spill context at calling place.

File: presto-main/src/main/java/com/facebook/presto/memory/AbstractAggregatedMemoryContext.java [hunk 5]
```diff
@@ -26,11 +28,21 @@
...
+        return () -> new AggregatedMemoryContext(this);
     }
 
     public LocalMemoryContext newLocalMemoryContext()
     {
-        return new LocalMemoryContext(this);
+        return localContextSupplier().get();
+    }
+
+    public Supplier<LocalMemoryContext> localContextSupplier()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/memory/AbstractAggregatedMemoryContext.java]: Why do we need `LocalMemoryContext` and `AggregatedMemoryContext`. Why not merge these two?

**Comment by findepi** [comment]
superseded by #8800

**Pull request closed without merging**